homeassistant/components/wemo.py

```diff
@@ -0,0 +1,108 @@
+"""
+homeassistant.components.wemo
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+WeMo device discovery.
+
+For more details about this component, please refer to the documentation at
+https://home-assistant.io/components/wemo/
+"""
+import logging
+
+from homeassistant import bootstrap
+from homeassistant.components import discovery
+from homeassistant.const import (
+    EVENT_HOMEASSISTANT_STOP, EVENT_PLATFORM_DISCOVERED,
+    ATTR_SERVICE, ATTR_DISCOVERED)
+
+REQUIREMENTS = ['pywemo==0.3.12']
+
+DOMAIN = 'wemo'
+DISCOVER_LIGHTS = 'wemo.light'
+DISCOVER_MOTION = 'wemo.motion'
+DISCOVER_SWITCHES = 'wemo.switch'
+
+# mapping from Wemo model_name to service
+WEMO_MODEL_DISPATCH = {
+    'Bridge':  DISCOVER_LIGHTS,
+    'Insight': DISCOVER_SWITCHES,
+    'Maker':   DISCOVER_SWITCHES,
+    'Motion':  DISCOVER_MOTION,
+    'Switch':  DISCOVER_SWITCHES,
+}
+
+_WEMO_SUBSCRIPTION_REGISTRY = None
+
+_LOGGER = logging.getLogger(__name__)
+
+
+# pylint: disable=unused-argument, too-many-function-args
+def setup(hass, config):
+    """ Common set up for WeMo devices """
+    import pywemo
+
+    global _WEMO_SUBSCRIPTION_REGISTRY
+    if _WEMO_SUBSCRIPTION_REGISTRY is None:
+        _WEMO_SUBSCRIPTION_REGISTRY = pywemo.SubscriptionRegistry()
+        _WEMO_SUBSCRIPTION_REGISTRY.start()
+
+        def stop_wemo(event):
+            """ Shutdown Wemo subscriptions and subscription thread on exit"""
+            _LOGGER.info("Shutting down subscriptions.")
+            _WEMO_SUBSCRIPTION_REGISTRY.stop()
+
+        hass.bus.listen_once(EVENT_HOMEASSISTANT_STOP, stop_wemo)
+
+    def discovery_event_listener(event):
+        """ Listens for discovery events. """
+        if event.data[ATTR_SERVICE] != discovery.SERVICE_WEMO:
+            return
+
+        # name, model, location, mac
+        discovery_info = event.data[ATTR_DISCOVERED]
+        model_name = discovery_info[1]
+        service = WEMO_MODEL_DISPATCH.get(model_name)
+
+        # do we know how to handle this service
+        if service is None:
+            return
+
+        hass.bus.fire(EVENT_PLATFORM_DISCOVERED, {
+            ATTR_SERVICE: service,
+            ATTR_DISCOVERED: discovery_info
+        })
+
+    hass.bus.listen(EVENT_PLATFORM_DISCOVERED, discovery_event_listener)
+
+    # Ensure components are loaded
+    bootstrap.setup_component(hass, 'light', config)
+    bootstrap.setup_component(hass, 'switch', config)
+
+    _LOGGER.info("Scanning for WeMo devices.")
+    devices = [(device.host, device) for device in pywemo.discover_devices()]
+
```

Comment:
I don't understand this part of HA but @balloob will know.

I think if you have discovery running this could discover the devices twice - once via `discovery_event_listener ` and once via `pywemo.discover_devices`.

I'm not sure how HA blocks this at present (but it doesn't run `pywemo.discover_devices` if you're using discovery) - so we should do the same here.